src/reduction_methods/set_cover.py

```diff
@@ -0,0 +1,83 @@
+import numpy as np
+
+from nptyping import Float, NDArray, Shape
+from tqdm import tqdm
+from typing import Optional
+
+from src.core import Database, memory, voxel_down_sample, VoxelGrid
+from src.loaders.providers import PointCloudProvider
+from src.reduction_methods.reduction_method import ReductionMethod
+
+
+@memory.cache
+def _get_voxel_by_frames_coverage_matrix(
+    voxel_grid: VoxelGrid,
+    pcds: list[PointCloudProvider],
+    poses: list[NDArray[Shape["4, 4"], Float]],
+):
+    voxels_enum = dict()
+    cur_voxel_num = 0
+    frames_voxels = []
+    for pcd_raw, pose in zip(pcds, poses):
+        pcd = pcd_raw.point_cloud.transform(pose)
+        down_sampled = voxel_down_sample(pcd, voxel_grid)
+        frame_voxels = []
+        for point in down_sampled.point.positions:
+            voxel_index = voxel_grid.get_voxel_index(point.numpy())
+            if voxel_index in voxels_enum:
+                num = voxels_enum[voxel_index]
+                frame_voxels.append(num)
+            else:
+                voxels_enum[voxel_index] = cur_voxel_num
+                frame_voxels.append(cur_voxel_num)
+                cur_voxel_num += 1
+        frames_voxels.append(frame_voxels)
+
+    coverage_mat = np.zeros((len(frames_voxels), len(voxels_enum)), dtype=np.uint8)
+
+    for i, voxels in enumerate(frames_voxels):
+        coverage_mat[i, voxels] = 1
+
+    return coverage_mat
+
+
+class SetCover(ReductionMethod):
+    """The method reduces database by set cover greedy algorithm"""
+
+    def __init__(self, db_size: Optional[int] = None, voxel_size: float = 0.1):
```

Comment:
db_size should be renamed, not clear what `size` means, it could be memory not amount of frames